marklogic-data-hub/src/main/java/com/marklogic/hub/hubcentral/migration/HubCentralMigrator.java

```diff
@@ -0,0 +1,167 @@
+     */
+    protected void migrateEntityModels() {
+        HubProject hubProject = hubConfig.getHubProject();
+        final File entityModelsDir = hubProject.getHubEntitiesDir().toFile();
+        if (!entityModelsDir.exists()) {
+            logger.warn("No entities directory exists, so no entity models will be migrated");
+            return;
+        }
+
+        logger.warn("Beginning migration of entity models in entities directory");
+
+        ObjectWriter writer = mapper.writerWithDefaultPrettyPrinter();
+        boolean atLeastOneEntityModelWasMigrated = false;
+
+        for (JsonNode entityModel : entityManager.getAllEntities()) {
+            ObjectNode entityModelNode = (ObjectNode) entityModel;
+
+            if (entityModelRequiresMigration(entityModelNode)) {
+                String title = entityModelNode.get("info").get("title").asText();
+                ObjectNode entityTypeNode = (ObjectNode) entityModelNode.get("definitions").get(title);
+
+                List<String> elementRangeIndex = mapper.convertValue(entityTypeNode.get("elementRangeIndex"), ArrayList.class);
+                elementRangeIndex = elementRangeIndex == null ? new ArrayList<>() : elementRangeIndex;
+                List<String> rangeIndex = mapper.convertValue(entityTypeNode.get("rangeIndex"), ArrayList.class);
+                rangeIndex = rangeIndex == null ? new ArrayList<>() : rangeIndex;
+                List<String> pathRangeIndex = mapper.convertValue(entityTypeNode.get("pathRangeIndex"), ArrayList.class);
+                pathRangeIndex = pathRangeIndex == null ? new ArrayList<>() : pathRangeIndex;
+                Set<String> mergedIndexArrays = new HashSet<>();
+                Stream.of(elementRangeIndex, rangeIndex, pathRangeIndex).forEach(mergedIndexArrays::addAll);
+
+                ObjectNode entityTypePropertiesNode = (ObjectNode) entityTypeNode.get("properties");
+                if (entityTypePropertiesNode == null) {
+                    logger.warn("entityTypePropertiesNode is null");
+                    entityTypeNode.remove(removableIndexArrays);
+                    atLeastOneEntityModelWasMigrated = true;
+                    continue;
+                }
+
+                entityTypePropertiesNode.fieldNames().forEachRemaining(propertyName -> {
+                    if (mergedIndexArrays.contains(propertyName)) {
+                        ObjectNode entityPropertyNode = (ObjectNode) entityTypePropertiesNode.get(propertyName);
+                        if (!isStructuredTypeProperty(entityPropertyNode)) {
+                            entityPropertyNode.put("facetable", true);
+                        }
+                    }
+                });
+                entityTypeNode.remove(removableIndexArrays);
+
+                File entityModelFile = Paths.get(hubProject.getHubEntitiesDir().toString(), title + entityManager.ENTITY_FILE_EXTENSION).toFile();
```

Comment:
This makes me nervous here, because we don't have a guarantee that this will overwrite the file that the entity was read from. The risk is that we write a new file and now the old file and migrated file both exist, and the user is then confused about what happened. 

For this reason, I think it's better to not use getAllEntities() from EntityManagerImpl. Instead, just do listFiles (and filter on name endsWith ".entity.json") on the entities dir and iterate over the list of files. That way, you know you're overwriting the correct File.